File: src/content/3.10/code/scala/snippet12.scala
def lambdaP[P[_, _]](P: Profunctor[P]): DiaProd[P] => ProdP[P] = {
  case DiaProd(paa) =>
    new ProdP[P] {
      def apply[A, B](f: A => B): P[A, B] =
        lambda(P)(paa[A])(f)
    }
}

def rhoP[P[_, _]](P: Profunctor[P]): DiaProd[P] => ProdP[P] = {
  case DiaProd(paa) =>
    new ProdP[P] {
      def apply[A, B](f: A => B): P[A, B] =
        rho(P)(paa[B])(f)
    }
}

File: src/content/3.10/code/ocaml/snippet12.ml
module EndsWithDiaProd
    (P : Profunctor)
    (D : DiaProd with type ('a, 'b) t = ('a, 'b) P.t)
    (PP : ProdP with type ('a, 'b) t = ('a, 'b) P.t) =
struct
  module E = EndsEqualizer (P)

  let lambdaP : 'a D.diaprod -> ('a, 'b) PP.prod_p =
   fun (DiaProd paa) -> E.lambda paa

  let rhoP : 'b D.diaprod -> ('a, 'b) PP.prod_p =
    fun (DiaProd pbb) -> E.rho pbb
end
